### PR TITLE
Build Kibana Docker Image Action

### PR DESCRIPTION
**Pull request by reakaleek**
## What does this PR do?

This PR creates an action to build and push kibana docker images.

Tested in 
- https://github.com/elastic/apm-pipeline-library/actions/workflows/test-kibana-docker-image-action.yml
- https://github.com/elastic/apm-pipeline-library/actions/workflows/test-build-and-publish-kibana-docker-image-action.yml

### CODE REVIEW

**Comment by kuisathaverat** [comment]
It probably fails, we are experiencing issues with building Kibana/main in the GitHub Action Runners, The issue looks like a OOM in Node.js.

File: .github/actions/build-and-publish-kibana-docker-image/action.yml [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+#    - name: Maximize build space
+#      uses: easimon/maximize-build-space@09ba4bd25af416c42e0142983065fcca855807a5
+
+    - uses: actions/checkout@v4
+      with:
+        repository: elastic/kibana
+        ref: ${{ inputs.ref }}
+        path: kibana-repo-checkout
+
+    - uses: elastic/apm-pipeline-library/.github/actions/docker-login@main
```
Comment (v1v) [inline .github/actions/build-and-publish-kibana-docker-image/action.yml:48]:
```suggestion
    - uses: elastic/apm-pipeline-library/.github/actions/docker-login@current
```

Comment (reakaleek) [reply .github/actions/build-and-publish-kibana-docker-image/action.yml]: https://github.com/elastic/apm-pipeline-library/pull/2348/commits/8ddaca66818a57646415d8b39ec2d4bfd09c04bb

File: .github/actions/build-and-publish-kibana-docker-image/action.yml [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+      if: steps.check-image-exists.outcome != 'success'
+      working-directory: kibana-repo-checkout
+      run: ${{ github.action_path }}/build.sh
+      shell: bash
+      env:
+        BUILD_SERVERLESS: ${{ inputs.serverless }}
+        DOCKER_NAMESPACE: ${{ steps.vars.outputs.docker-namespace }}
+        DOCKER_TAG: ${{ steps.vars.outputs.docker-tag }}
+
+    - uses: elastic/apm-pipeline-library/.github/actions/docker-login@main
```
[the remark this answers is not in the document]
Comment (reakaleek) [reply .github/actions/build-and-publish-kibana-docker-image/action.yml]: https://github.com/elastic/apm-pipeline-library/pull/2348/commits/8ddaca66818a57646415d8b39ec2d4bfd09c04bb

File: .github/actions/kibana-docker-image/README.md [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+| `git-ref`                               | String | no       | The git ref of the repository. (Default: default branch, e.g. `main`)     |
+| `serverless`                            | String | no       | Whether to build serverless images or not. (Default: `false`)             |
+| `docker-registry`                       | String | no       | The docker registry for pushing the image. (Default: `docker.elastic.co`) |
+| `docker-namespace`                      | String | no       | The namespace of the repository. (Default: `observability-ci`)            |
+
+## Outputs
+
+| Name   | Description                                         |
+|--------|-----------------------------------------------------|
+| `ref`  | The published docker image reference. (`image:tag`) |
```
Comment (v1v) [inline .github/actions/kibana-docker-image/README.md:57]: neat!

File: .github/actions/kibana-docker-image/README.md [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+```
+
+## Inputs
+
+Following inputs can be used as `step.with` keys
+
+| Name                                    | Type   | Required | Description                                                               |
+|-----------------------------------------|--------|----------|---------------------------------------------------------------------------|
+| `git-repository`                        | String | no       | The git repository to checkout. (Default: `elastic/kibana`)               |
```
Comment (v1v) [inline .github/actions/kibana-docker-image/README.md:47]: **nit**: github-repository instead?

File: .github/actions/kibana-docker-image/action.yml [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+    - uses: actions/setup-node@v3
+      with:
+        node-version-file: kibana-repo-checkout/.node-version
+        cache: yarn
+        cache-dependency-path: kibana-repo-checkout/yarn.lock
+
+    - name: Set up variables
+      id: vars
+      working-directory: kibana-repo-checkout
+      run: |
```
Comment (v1v) [inline .github/actions/kibana-docker-image/action.yml:48]: **nit**: what do you think moving this part to a script?

File: .github/actions/kibana-docker-image/action.yml [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+    value: ${{ steps.vars.outputs.docker-reference }}
+
+runs:
+  using: composite
+  steps:
+    - uses: actions/checkout@v4
+      with:
+        repository: ${{ inputs.git-repository }}
+        ref: ${{ inputs.git-ref }}
+        path: kibana-repo-checkout
```
Comment (v1v) [inline .github/actions/kibana-docker-image/action.yml:37]: **nit**: what do you think using the toplevel env: CHECKOUT_DIR: "kibana-repo-checkout"

then


```suggestion
        path: ${{ env.CHECKOUT_DIR }}
```

and in the other places too?

File: .github/workflows/test-kibana-docker-image-action.yml [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+on:
+  workflow_dispatch: ~
+  push:
+    branches:
+      - main
+    paths:
+      - ".github/actions/kibana-docker-image/**"
+  schedule:
+    - cron: '0 6 * * *'
```
Comment (v1v) [inline .github/workflows/test-kibana-docker-image-action.yml:15]: That's great! So we can discover when things get broken

File: .github/workflows/test-kibana-docker-image-action.yml [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+jobs:
+  kibana-cloud-docker-image:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: elastic/apm-pipeline-library/.github/actions/docker-login@current
```
Comment (v1v) [inline .github/workflows/test-kibana-docker-image-action.yml:22]: **nit**: GitHub workflows in this particular project can use `main` so we can detect when things get broken before we even cut a new release.

```suggestion
      - uses: elastic/apm-pipeline-library/.github/actions/docker-login@main
```

File: .github/workflows/test-kibana-docker-image-action.yml [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+      - uses: ./.github/actions/kibana-docker-image
+        id: kibana-docker-image
+      - name: Verify
+        run: |
+          echo "${DOCKER_IMAGE_REF:?}"
+          docker pull "${DOCKER_IMAGE_REF}"
+        env:
+          DOCKER_IMAGE_REF: ${{ steps.kibana-docker-image.outputs.ref }}
+      - if: failure()
+        uses: elastic/apm-pipeline-library/.github/actions/notify-build-status@current
```
Comment (v1v) [inline .github/workflows/test-kibana-docker-image-action.yml:38]: ditto:

```suggestion
        uses: elastic/apm-pipeline-library/.github/actions/notify-build-status@main
```

File: .github/workflows/test-kibana-docker-image-action.yml [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+          vaultUrl: ${{ secrets.VAULT_ADDR }}
+          vaultRoleId: ${{ secrets.VAULT_ROLE_ID }}
+          vaultSecretId: ${{ secrets.VAULT_SECRET_ID }}
+          slackChannel: "#observablt-bots"
+
+  kibana-serverless-docker-image:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: elastic/apm-pipeline-library/.github/actions/docker-login@current
```
Comment (v1v) [inline .github/workflows/test-kibana-docker-image-action.yml:49]:
```suggestion
      - uses: elastic/apm-pipeline-library/.github/actions/docker-login@main
```

File: .github/workflows/test-kibana-docker-image-action.yml [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+        with:
+          serverless: true
+      - name: Verify
+        run: |
+          echo "${DOCKER_IMAGE_REF:?}"
+          docker pull "${DOCKER_IMAGE_REF}"
+        env:
+          DOCKER_IMAGE_REF: ${{ steps.kibana-docker-image.outputs.ref }}
+      - if: failure()
+        uses: elastic/apm-pipeline-library/.github/actions/notify-build-status@current
```
Comment (v1v) [inline .github/workflows/test-kibana-docker-image-action.yml:67]:
```suggestion
        uses: elastic/apm-pipeline-library/.github/actions/notify-build-status@main
```

**Review by v1v — DISMISSED** [review]
LGTM with just a few comments

File: .github/actions/kibana-docker-image/action.yml (hunk 6 above)
Comment (reakaleek) [reply .github/actions/kibana-docker-image/action.yml]: https://github.com/elastic/apm-pipeline-library/pull/2348/commits/3deb3b43a6d5e21adb6724815fe16281f6a3cc65

File: .github/actions/kibana-docker-image/README.md (hunk 4 above)
Comment (reakaleek) [reply .github/actions/kibana-docker-image/README.md]: https://github.com/elastic/apm-pipeline-library/pull/2348/commits/e81e04169d4614c878128a4feb4c81417c3e78ac

File: .github/workflows/test-kibana-docker-image-action.yml (hunk 11 above)
Comment (reakaleek) [reply .github/workflows/test-kibana-docker-image-action.yml]: https://github.com/elastic/apm-pipeline-library/pull/2348/commits/3eddef7c034fecd2fb8970ab07dcb5873a89b956

File: .github/workflows/test-kibana-docker-image-action.yml (hunk 10 above)
Comment (reakaleek) [reply .github/workflows/test-kibana-docker-image-action.yml]: https://github.com/elastic/apm-pipeline-library/pull/2348/commits/3eddef7c034fecd2fb8970ab07dcb5873a89b956

File: .github/workflows/test-kibana-docker-image-action.yml (hunk 9 above)
Comment (reakaleek) [reply .github/workflows/test-kibana-docker-image-action.yml]: https://github.com/elastic/apm-pipeline-library/pull/2348/commits/3eddef7c034fecd2fb8970ab07dcb5873a89b956

File: .github/actions/kibana-docker-image/action.yml (hunk 5 above)
Comment (reakaleek) [reply .github/actions/kibana-docker-image/action.yml]: https://github.com/elastic/apm-pipeline-library/pull/2348/commits/f1a719deb787b8593e57e8c456a2b03783bc1a86

File: .github/actions/kibana-docker-image/build-and-push.sh [hunk 12]
```diff
@@ -0,0 +1,52 @@
+#!/usr/bin/env bash
```
Comment (v1v) [inline .github/actions/kibana-docker-image/build-and-push.sh:1]: I cannot see how the push works

File: .github/actions/kibana-docker-image/build-and-push.sh [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+if [ "${SERVERLESS}" == "false" ] ; then
+  skip_docker_flag="--skip-docker-serverless"
+else
+  skip_docker_flag="--skip-docker-cloud"
+fi
+time node scripts/build \
+      --docker-images \
+      --docker-namespace="${DOCKER_NAMESPACE}" \
+      --docker-tag="${DOCKER_TAG}" \
+      --docker-push \
```
Comment (reakaleek) [inline .github/actions/kibana-docker-image/build-and-push.sh:43]: @v1v Here is the push option.

Comment (v1v) [reply .github/actions/kibana-docker-image/build-and-push.sh:43]: ahhh!! I missed that flag, nice! Thanks 🙏

**Review by v1v — APPROVED** [review]

**Pull request merged**